(feat) Introduce type-safe colour-scheme abstraction for theming (V100)

tobitege opened the pull request:
### Summary  
This PR introduces a type-safe colour-scheme architecture and migrates the **Microsoft 365 Black** palette as the working example.

### Key changes  
1. **New abstraction**  
   * `AbstractBaseColorScheme` – one property per `SchemeBaseColors` entry.  
   * `ArrayBaseColorScheme` – adapter so existing colour arrays still work.  
   * Overloads in `PaletteMicrosoft365Base` accept either adapter or any concrete scheme.

2. **Example migration**  
   * Added `PaletteMicrosoft365BlackScheme` (property class).  
   * `PaletteMicrosoft365Black` now consumes the scheme object; the 15 KB `_schemeBaseColors` array is gone.

3. **Tooling & docs**  
   * `Scripts/generate_scheme_classes.py` – auto-generates *_Scheme.cs files; supports `--single` and `--dry-run`.  
   * Comprehensive README in *Base* folder explaining concept, migration path and benefits.

### About the helper script – `Scripts/generate_scheme_classes.py`

A new Python utility is included to automate the creation of strongly-typed *Scheme* classes and make the migration virtually effortless.

Key points  
• **One-shot array migration** – running the script from the repo's Scripts folder:
```bash  
python Scripts/generate_scheme_classes.py  
```  
scans every palette that still contains `_schemeBaseColors`, parses the colour literals, and emits a matching `PaletteXXXXScheme.cs` file next to it. With ~33 themes it finishes in a couple of seconds on a standard laptop.

• **Idempotent & safe** – if a *_Scheme.cs* file already exists the script prints  
  `Skip <FileName> (already exists)` and leaves the file untouched, so it can be rerun at any time without risking local edits.

• **Selective generation** –  
  * `--single`  generates the first missing scheme only (useful for step-by-step PRs).  
  * `--dry-run`  shows what would be generated without writing any files.

• **Complete coverage** – the script walks the ​`SchemeBaseColors`​ enumeration, pads missing colours with `GlobalStaticValues.EMPTY_COLOR`, and guarantees every property is present, so the resulting class compiles out-of-the-box.

With this tool, converting all remaining palettes is literally a matter of running the script and swapping the constructor argument to the new scheme instance—no hand-typing of 230+ properties per theme required.

### Benefits  
* **Compile-time safety** – missing or mis-ordered colours now surface as compiler errors.  
* **Readable code-reviews** – colours referenced by name, not magic indexes.  
* **Zero runtime impact** – rendering still receives the same `Color[]`.  
* **Incremental adoption** – legacy palettes keep working; new ones can be migrated in minutes with the script.

tobitege commented:
Thanks, changes applied.

tobitege commented:
- Removed `ArrayBaseColorScheme.cs` completely
- The abstract class now has writable auto-properties for all colors
- Enhanced py script with more command line args, detects missing values
- Updated README.md accordingly
- Regenerated the *365BlackScheme.cs file with current script

@giduac thanks for the feedback, I think this cleans it up well and makes this a real contender, especially since a palette conversion is 95% automatic now (at least for the SchemeBaseColors-based ones).

tobitege commented:
Added color editing functions and updated README accordingly.

Added F6 color picker in PaletteViewer to live edit theme values! 
Load PaletteMicrosoft365Black palette and make it the current theme, then click a cell, press F6, choose color. 

<img width="946" height="325" alt="{56014C8D-8966-45CB-A17A-D08BE5E21584}" src="https://github.com/user-attachments/assets/277d7a84-0532-4b3d-bb8a-60245519abd4" />

PWagner1 commented on Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/README.md:
```diff
@@ -0,0 +1,256 @@
+# Color-Scheme Abstraction for Krypton Toolkit
```
Can this also be included in here please? [Krypton-Suite/Standard-Toolkit-Online-Help](https://github.com/Krypton-Suite/Standard-Toolkit-Online-Help)

PWagner1 reviewed (commented):
Please see comments

giduac commented:
**First try for a TODO List.**
Probably all of this will need a bit more discussion but we will update this table until ready to go.

- A list of themes the script will be editing
- A full array comment block that will be detected by the script as being correct.

- The current themes need to get a check each to see if it's visually working correct.
This is ONLY about the themes the script will be editing (not those that are skipped by the script).
Because if a theme is working correctly then inserting "missing items" will put it off.
These sort of themes should be manually be updated before the script is run. 
If such a theme does not have the required 232 items add items with GlobalStaticValues.EMPTY_COLOR
Then replace the comment block with our new standard comment block that will be correctly detected by the script so that the script does update the array.
Also the script can extract the theme colours to the new class type.

- Discuss and come to a naming convention for the classes derived from KryptonColorSchemeBase.
The name in 2294 is: PaletteMicrosoft365BlackScheme.cs

Idea behind this to use this class format for the other colour arrays as well resulting in for example:
PaletteMicrosoft365Black_BaseScheme
PaletteMicrosoft365Black_TrackBarScheme
etc.

- Classes derived from KryptonColorSchemeBase will be sealed to avoid a potential new slew of colour scheme complications through a hierarchy of the colour schemes.
- 
- Discuss how/where are the new files stored in the directory structure
Would a grouping by theme be an option:

```
- PaletteMicrosoft365Black
 |  - PaletteMicrosoft365Black_BaseScheme
 |  - PaletteMicrosoft365Black_Images
```

- Remove from `PaletteMicrosoft365Base` `private static Color[] ConvertSchemeToArray(KryptonColorSchemeBase scheme)`
- Rename AbstractBaseColorScheme to KryptonColorSchemeBase

- Add ToArray() method to KryptonColorSchemeBase
```cs

	#region Methods
	public Color[] ToArray()
	{
		var names = Enum.GetNames(typeof(SchemeBaseColors));
		var colors = new Color[names.Length];
		var type = this.GetType();

		for (int i = 0; i < names.Length; i++)
		{
			colors[i] = type.GetProperty(names[i])?.GetValue(this) is Color color
				? color
				: GlobalStaticValues.EMPTY_COLOR;
		}
		return colors;
	}
	#endregion
```	

- In the base palette add
`private readonly KryptonColorSchemeBase _colorScheme;`

- Change ctor in base palette to:
```cs
    /// <summary>
    /// Overload that accepts any AbstractBaseColorScheme implementation.
    /// Converts it to a Color[] and forwards to the main constructor.
    /// </summary>
    protected PaletteMicrosoft365Base(
        KryptonColorSchemeBase scheme,
        ImageList checkBoxList,
        ImageList galleryButtonList,
        Image?[]   radioButtonArray,
        Color[]    trackBarColors)
        : this(scheme.ToArray(),
               checkBoxList,
               galleryButtonList,
               radioButtonArray,
               trackBarColours)
    {
        _colorScheme = scheme;
    }
```

tobitege commented:
Refactored scheme generation:
- Renamed AbstractBaseColorScheme to KryptonColorSchemeBase and related references. 
- Introducing class KryptonColorTrackBarSchemeBase
- Introducing new color enum SchemeTrackBarColors for track bars.
- Enhance SchemeGenerator to also extract _trackBarColors for SchemeTrackBarColors
- Generated files named *_BaseScheme.cs and *_TrackBarScheme.cs
- Added private _colorScheme to PaletteMicrosoft365Base
- Adapted PaletteMicrosoft365Base ctor to store colors in _colorScheme
- Removed color indexer from PaletteMicrosoft365Base
- Removed ConvertSchemeToArray from PaletteMicrosoft365Base
- Moved ToArray from PaletteMicrosoft365Base into SchemeBaseColorsExtensions.cs as extension on KryptonColorSchemeBase
- Split several themes into separate base and theme files so later automation can work with *Base.cs filenames
- kptheme tool now only ends after final user input

Separate points:

The `kptheme` by default creates files in the same folder as the source palette file itself.
But is has an arg to specify an output folder for all generated file(s) to go to (in the current call).

Source palette files are NOT touched by the scheme generator, i.e. no automatic removal of arrays.
This should happen on a per-palette-file basis (removal of arrays, ctor adaptations etc).

Open:
- [ ] First 3 bullet points from the todo
- [ ] `PaletteMicrosoft365Black.cs` is the only one thus far with the 2 generated files (trackbar class not wired yet)

**Any further changes to palette files should happen in followup PR's as it already is a lot to review!**

tobitege commented:
Bugfix in scheme generator, fixed PaletteMicrosoft365White_BaseScheme and PaletteMicrosoft365Black_BaseScheme files.

PWagner1 commented on Directory.Build.props:
```diff
@@ -83,6 +83,10 @@
 	</Choose>
 
+	<PropertyGroup Condition="'$(TargetFramework)' == 'net10.0-windows'">
+		<GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>
```
Just curious, what does this do?

tobitege replied:
_With the newer SDK toolset a number of the types referenced in older WinForms .resx files (System.Drawing types, design-time helpers, etc.) are either missing, moved or only available in compatibility shims. When ResGen tries to spin up those types in a net10 build it fails, so the build stops with MSB3xxx / CSxxxx errors.
The fix
Setting
`<GenerateResourceUsePreserializedResources>true</GenerateResourceUsePreserializedResources>`
tells ResGen: “If the .resx already contains a pre-serialised binary blob, just copy it; don’t try to re-hydrate the objects.” Visual Studio’s designer always stores that pre-serialised payload, so nothing is lost. Wrapping it in the condition
Condition="'$(TargetFramework)' == 'net10.0-windows'"
limits the change to that TF where the problem occurs, keeping the default behaviour for the other targets.
Bottom line: the property sidesteps the failing re-serialisation step, letting the net10.0-windows build succeed without touching the resources themselves._

PWagner1 commented on Source/Krypton Components/Krypton.Toolkit/Utilities/SchemeBaseColorsExtensions.cs:
```diff
@@ -0,0 +1,58 @@
+using System.Drawing;
+
```
License needed + move `using System.Drawing;` -> `GlobalDeclarations.cs`

PWagner1 reviewed (commented):
Please see comments

tobitege commented:
<img width="888" height="363" alt="image" src="https://github.com/user-attachments/assets/fdf51750-d460-4b64-9359-6f3eeab1cdce" />

Changelog for the commits since yesterday will follow later today.

tobitege commented:
One note: BEFORE merge, I'll remove the interim Python scripts, that are not needed/used anymore, as well as the `2021-prd-V100.md` document.
The newly added README.md should suffice to explain the mechanics of this PR.

PWagner1 commented:
No other comments from me

tobitege commented:
> @tobitege
> 
> Can I contribute some changes? :)

What kind of changes do you have in mind? Anything that could be done in a followup PR?

PWagner1 commented:
> > @tobitege
> > Can I contribute some changes? :)
> 
> What kind of changes do you have in mind? Anything that could be done in a followup PR?

Inline documentation, nothing major...

tobitege commented:
> > > @tobitege
> > > Can I contribute some changes? :)
> > 
> > 
> > What kind of changes do you have in mind? Anything that could be done in a followup PR?
> 
> Inline documentation, nothing major...

Lets do that after the merge, please, or if you have general directions which kind and where, I could give that a stab. But I'd rather want to get this done asap (time constraints).

tobitege commented:
@Smurf-IV @PWagner1 @giduac 
If you guys prefer, I can remove the extra 2 "helper" projects that helped during the conversion of all the themes (`kptheme`, `Krypton.ThemeGen`).
If they ever could be useful once more, I could resubmit them.

tobitege commented:
> > @Smurf-IV @PWagner1 @giduac If you guys prefer, I can remove the extra 2 "helper" projects that helped during the conversion of all the themes (`kptheme`, `Krypton.ThemeGen`). If they ever could be useful once more, I could resubmit them.
> 
> @tobitege
> 
> You could always 'unload' them?

It's more that they're now without concrete use and to end-users they'd just be a confusing and if used wrongly, could change source files. 
Yeh, I'll just move them out and keep them as backup.
